packages-exp/auth-exp/src/platform_browser/recaptcha/recaptcha_loader.ts

```diff
@@ -0,0 +1,139 @@
+}
+
+function pullGrecaptcha(): Recaptcha | undefined {
+  return (window as AuthWindow).grecaptcha;
+}
+
+const WINDOW = window as AuthWindow;
+
+/**
+ * Loader for the GReCaptcha library. There should only ever be one of this.
+ */
+export class ReCaptchaLoaderImpl implements ReCaptchaLoader {
+  private hl = '';
+  private counter = 0;
+  private readonly librarySeparatelyLoaded = !!pullGrecaptcha();
+
+  load(auth: Auth, hl = ''): Promise<Recaptcha> {
+    if (this.shouldResolveImmediately(hl)) {
+      return Promise.resolve(pullGrecaptcha()!);
+    }
+    return new Promise<Recaptcha>((resolve, reject) => {
+      const networkTimeout = WINDOW.setTimeout(() => {
+        reject(
+          AUTH_ERROR_FACTORY.create(AuthErrorCode.NETWORK_REQUEST_FAILED, {
+            appName: auth.name
+          })
+        );
+      }, NETWORK_TIMEOUT_DELAY.get());
+
+      WINDOW[_JSLOAD_CALLBACK] = () => {
+        WINDOW.clearTimeout(networkTimeout);
+        delete WINDOW[_JSLOAD_CALLBACK];
+
+        const recaptcha = pullGrecaptcha();
+
+        if (!recaptcha) {
+          reject(
+            AUTH_ERROR_FACTORY.create(AuthErrorCode.INTERNAL_ERROR, {
+              appName: auth.name
+            })
+          );
+          return;
+        }
+
+        // Wrap the greptcha render function so that we know if the developer has
+        // called it separately
+        const render = recaptcha.render;
+        recaptcha.render = (container, params) => {
+          const widgetId = render(container, params);
+          this.counter++;
+          return widgetId;
+        };
+
+        this.hl = hl;
+        resolve(recaptcha);
+      };
+
+      const url = `${RECAPTCHA_BASE}?${querystring({
+        onload: _JSLOAD_CALLBACK,
+        render: 'explicit',
+        hl
+      })}`;
+
+      jsHelpers._loadJS(url).catch(() => {
+        clearTimeout(networkTimeout);
+        reject(
+          AUTH_ERROR_FACTORY.create(AuthErrorCode.INTERNAL_ERROR, {
+            appName: auth.name
+          })
+        );
+      });
+    });
+  }
+
+  clearedOneInstance(): void {
+    this.counter--;
+  }
+
+  private shouldResolveImmediately(hl: string): boolean {
+    // We can resolve immediately if:
+    //   • grecaptcha is already defined AND (
+    //     1. the requested language codes are the same OR
+    //     2. there exists already a ReCaptcha on the page
+    //     3. the library was already loaded by the app
+    // In cases (2) and (3), we _can't_ reload as it would break the recaptchas
+    // that are already in the page
+    return (
+      !!pullGrecaptcha() &&
+      (hl === this.hl || this.counter > 0 || this.librarySeparatelyLoaded)
+    );
+  }
+}
+
+class MockReCaptchaLoaderImpl implements ReCaptchaLoader {
+  async load(auth: Auth): Promise<Recaptcha> {
+    return new MockReCaptcha(auth);
+  }
+
+  clearedOneInstance(): void {}
+}
+
+export const MOCK_RECAPTCHA_LOADER: ReCaptchaLoader = new MockReCaptchaLoaderImpl();
```

Comment:
should the mock belong in test/ instead of src/?

Reply:
No, this isn't a mock for library testing purposes, it's a mock for `appVerificationDisabledForTesting` purposes. This is used rather than the real recaptcha for doing SMS login using fake numbers